aggregator/src/test/java/org/jenkinsci/plugins/workflow/steps/SynchronousNonBlockingStepTest.java

```diff
@@ -0,0 +1,245 @@
+        SynchronousNonBlockingStep.notify("wait");
+
+        System.out.println("Waiting until syncnonblocking (and the full flow) finishes");
+        j.waitForCompletion(b);
+        System.out.println("Build finished. Continue.");
+        // Check for the last message
+        j.assertLogContains("Second message", b);
+        j.assertBuildStatusSuccess(b);
+    }
+
+    @Test
+    public void interruptedTest() throws Exception {
+        WorkflowJob p = j.jenkins.createProject(WorkflowJob.class, "p");
+        p.setDefinition(new CpsFlowDefinition("node {\n" +
+            "echo 'First message'\n" +
+            "syncnonblocking 'waitinterrupt'\n" +
+            "echo 'Second message'\n" +
+        "}"));
+        WorkflowRun b = p.scheduleBuild2(0).getStartCondition().get();
+
+        // Wait for syncnonblocking to be started
+        System.out.println("Waiting to syncnonblocking to start...");
+        // waitinterrupt is a special key that makes syncnonblocking step to wait inside a while  loop checking
+        // for Thread.interrupted()
+        SynchronousNonBlockingStep.waitForStart("waitinterrupt", b);
+
+        // At this point syncnonblocking is looping and checking for interruption
+
+        CpsFlowExecution e = (CpsFlowExecution) b.getExecutionPromise().get();
+        // Let's force a call to stop. This will try to send an interruption to the run Thread
+        e.interrupt(Result.ABORTED);
+        System.out.println("Looking for interruption received log message");
+        j.waitForMessage("Run thread interrupted", b);
+        j.waitForCompletion(b);
+    }
+
+    @Test
+    public void parallelTest() throws Exception {
+        WorkflowJob p = j.jenkins.createProject(WorkflowJob.class, "p");
+        p.setDefinition(new CpsFlowDefinition("node {\n" +
+            "echo 'First message'\n" +
+            "parallel( a: { syncnonblocking 'wait0'; echo 'a branch'; }, b: { echo 'b branch' } )\n" +
+            "echo 'Second message'\n" +
+        "}"));
+        WorkflowRun b = p.scheduleBuild2(0).getStartCondition().get();
+
+        SynchronousNonBlockingStep.waitForStart("wait0", b);
+
+        // Wait for "b" branch to print its message
+        j.waitForMessage("b branch", b);
+        System.out.println("b branch finishes");
+
+        // Check that "a" branch is effectively blocked
+        j.assertLogNotContains("a branch", b);
+
+        // Notify "a" branch
+        System.out.println("Continue on wait0");
+        SynchronousNonBlockingStep.notify("wait0");
+
+        // Wait for "a" branch to finish
+        j.waitForMessage("a branch", b);
+        j.waitForCompletion(b);
+    }
+
+    public static final class SynchronousNonBlockingStep extends AbstractStepImpl implements Serializable {
+
+        public static final class State {
+            private static State state;
+            static synchronized State get() {
+                if (state == null) {
+                    state = new State();
```

Comment:
This works well enough here, where we are using a `@TestExtension` and so the set of callers is strictly limited and you can ensure that every test uses a distinct `id`, but FYI if you want to make `State.get()` be independent from test run to test run, you would need to tie the `State` instance to the `Jenkins` instance. Most easily by making it a non-`static` field in `DescriptorImpl`.